go/array_string/min_window/solution.go

```diff
@@ -0,0 +1,101 @@
+// package main has the solution for the minimum window
+// algorithm implementation using the sliding window technique.
+package main
+
+import (
+	"fmt"
+	"math"
+	"strings"
+)
+
+func minWindow(mainStr, subStr string) string {
+	// validation
+	if len(mainStr) == 0 || len(subStr) == 0 || len(subStr) > len(mainStr) {
+		return ""
+	}
+	// using the sliding window technique.
+	// create hashmap for required and desired character.
+	required := make(map[string]int)
+	desired := make(map[string]int)
+	// ensure it is the count of the unique keys in desired
+	// and not the subStr incase of repeated characters.
+	substrCount := 0
+
+	// populate the desired and required hashmaps
+	for _, char := range subStr {
+		desired[string(char)] = 0
+
+		// create the new hashmap with the count of the required
+		// characters in the subStr.
+		if _, ok := required[string(char)]; !ok {
+			// if the character has not been accounted of.
+			substrCount++ // count the unique characters
+			required[string(char)] = strings.Count(subStr, string(char))
+		}
+	}
+
+	var slowPointer, fastPointer uint64
+	minWindowRange := []uint64{0, math.MaxUint64}
+
+	// enlarge the window to get the desirable character range
+	for fastPointer < uint64(len(mainStr)) {
+		// check if there is range match
+		char := string(mainStr[fastPointer])
+
+		if _, ok := desired[char]; ok {
+			desired[char]++
+
+			// check if the window contains the matching required characters
+			desiredValue := desired[char]
+			requiredValue := required[char]
+			if desiredValue == requiredValue {
```

Comment:
`if desired[char] == required[char] {`

Defining a new variable means that someone reading your code has to store it in their head until it is used. We only have so much RAM in our heads, and our processor cache is even smaller so there's no point using it up :stuck_out_tongue_closed_eyes:

Reply:
@aschlosberg Lool, I totally understand. :laughing: (the part of the processor cache of human brain cracked me up).